File: frontend/provas/prova_razao_social.js
// Prova de frontend/src/pages/razaoSocial.js — Node puro, sem build.
//   node frontend/provas/prova_razao_social.js

import { formatarRazaoSocial as f } from '../src/pages/razaoSocial.js';

let ok = 0, falhou = 0;
const eq = (entrada, esperado) => {
  const obtido = f(entrada);
  if (obtido === esperado) { ok++; console.log(`  ok  ${JSON.stringify(entrada)} -> ${JSON.stringify(obtido)}`); }
  else { falhou++; console.log(`FALHA  ${JSON.stringify(entrada)}\n       obtido:   ${JSON.stringify(obtido)}\n       esperado: ${JSON.stringify(esperado)}`); }
};

console.log('\n1) Os três nomes do print, que era o caso concreto');
eq('Mark Building Gerenc. Predial Ltda.', 'Mark Building Gerenc. Predial Ltda.');
eq('RIO BRAVO COM. DE ARMAS MUNIÇÕES E ACESS. LTDA', 'Rio Bravo Com. de Armas Munições e Acess. Ltda');
eq('TROPS CENTRO DE ESP. E LAZER LTDA-ME', 'Trops Centro de Esp. e Lazer Ltda-ME');

console.log('\n2) Quem já tem minúscula fica intacto — só apara espaço sobrando');
eq('  Alpha   Serviços Ltda  ', 'Alpha Serviços Ltda');
eq('CW Administra Ltda.', 'CW Administra Ltda.');

console.log('\n3) Sigla do grupo continua sigla, palavra curta não vira sigla');
eq('MKB PARTICIPACOES LTDA', 'MKB Participacoes Ltda');
eq('IWC COMERCIO LTDA', 'IWC Comercio Ltda');
eq('RIO NEGRO TRANSPORTES', 'Rio Negro Transportes');
eq('CW ADMINISTRA LTDA', 'CW Administra Ltda');
eq('BPS4 OUTSOURCING LTDA', 'BPS4 Outsourcing Ltda');

console.log('\n4) Átona só perde a maiúscula longe do começo');
eq('DE PAULA E FILHOS LTDA', 'De Paula e Filhos Ltda');
eq('CASA DAS TINTAS DO NORTE', 'Casa das Tintas do Norte');

console.log('\n5) Forma jurídica com grafia própria');
eq('COMERCIAL XPTO S/A', 'Comercial Xpto S.A.');
eq('PADARIA CENTRAL EIRELI', 'Padaria Central Eireli');
eq('GRAFICA SOL LTDA-EPP', 'Grafica Sol Ltda-EPP');

console.log('\n6) Abreviação com ponto é palavra encurtada, não sigla');
eq('ESP. E LAZER', 'Esp. e Lazer');
eq('IND. E COM. DE PECAS', 'Ind. e Com. de Pecas');

console.log('\n7) Vazio e nulo não quebram o cabeçalho');
eq('', ''); eq(null, ''); eq(undefined, ''); eq('   ', '');

console.log(`\n${falhou === 0 ? 'TUDO VERDE' : 'VERMELHO'} — ${ok} ok, ${falhou} falhou\n`);
process.exit(falhou === 0 ? 0 : 1);

File: backend/provas/prova_razao_social.py
"""
prova_razao_social.py — a razão social escrita igual nos dois lados.

O mesmo algoritmo existe no backend (mensagem ao cliente) e no frontend (telas).
Esta prova roda os dois nos MESMOS casos e compara: sem isso, uma correção num
lado passaria despercebida no outro, e o cliente veria o nome de um jeito na
tela e de outro no WhatsApp.

    python provas/prova_razao_social.py
"""
from __future__ import annotations

import json
import os
import shutil
import subprocess
import sys
from pathlib import Path

RAIZ = Path(__file__).resolve().parent.parent
sys.path.insert(0, str(RAIZ))
os.environ.setdefault("DATABASE_URL", "sqlite:///:memory:")

from app.services.razao_social import formatar   # noqa: E402

ok = True
def checa(nome, cond, extra=""):
    global ok
    print(("  ok   " if cond else "FALHA  ") + nome + (f"  {extra}" if extra else ""))
    ok = ok and bool(cond)


CASOS = [
    ("Mark Building Gerenc. Predial Ltda.", "Mark Building Gerenc. Predial Ltda."),
    ("RIO BRAVO COM. DE ARMAS MUNIÇÕES E ACESS. LTDA", "Rio Bravo Com. de Armas Munições e Acess. Ltda"),
    ("TROPS CENTRO DE ESP. E LAZER LTDA-ME", "Trops Centro de Esp. e Lazer Ltda-ME"),
    ("MKB PARTICIPACOES LTDA", "MKB Participacoes Ltda"),
    ("IWC COMERCIO LTDA", "IWC Comercio Ltda"),
    ("RIO NEGRO TRANSPORTES", "Rio Negro Transportes"),
    ("CW ADMINISTRA LTDA", "CW Administra Ltda"),
    ("BPS4 OUTSOURCING LTDA", "BPS4 Outsourcing Ltda"),
    ("DE PAULA E FILHOS LTDA", "De Paula e Filhos Ltda"),
    ("CASA DAS TINTAS DO NORTE", "Casa das Tintas do Norte"),
    ("COMERCIAL XPTO S/A", "Comercial Xpto S.A."),
    ("PADARIA CENTRAL EIRELI", "Padaria Central Eireli"),
    ("GRAFICA SOL LTDA-EPP", "Grafica Sol Ltda-EPP"),
    ("IND. E COM. DE PECAS", "Ind. e Com. de Pecas"),
    ("  Alpha   Serviços Ltda  ", "Alpha Serviços Ltda"),
    ("", ""),
]

print("\n1. O backend escreve como se espera")
for entrada, esperado in CASOS:
    checa(f"{entrada!r} -> {esperado!r}", formatar(entrada) == esperado,
          "" if formatar(entrada) == esperado else f"(deu {formatar(entrada)!r})")
checa("nulo não quebra", formatar(None) == "")

print("\n2. E escreve IGUAL ao frontend")
node = shutil.which("node")
if not node:
    print("  --   node não encontrado; comparação entre os lados pulada")
else:
    # Caminho RELATIVO ao .mjs: import ESM com caminho absoluto exigiria
    # `file://`, e o erro que ele dá não diz isso.
    script = (
        "import {formatarRazaoSocial as f} from "
        "'../../frontend/src/pages/razaoSocial.js';\n"
        "const casos=JSON.parse(process.argv[2]);\n"
        "console.log(JSON.stringify(casos.map(c=>f(c))));\n"
    )
    entradas = [c[0] for c in CASOS]
    arq = RAIZ / "provas" / "_cmp_razao.mjs"
    arq.write_text(script, encoding="utf-8")
    try:
        saida = subprocess.run([node, str(arq), json.dumps(entradas)],
                               capture_output=True, text=True, timeout=30)
        if saida.returncode != 0:
            checa("o frontend rodou", False, saida.stderr.strip()[:120])
        else:
            do_front = json.loads(saida.stdout)
            do_back = [formatar(e) for e in entradas]
            divergentes = [(e, b, f) for e, b, f in zip(entradas, do_back, do_front) if b != f]
            checa("os dois lados escrevem igual em todos os casos",
                  not divergentes,
                  "" if not divergentes else "; ".join(
                      f"{e!r}: back={b!r} front={f!r}" for e, b, f in divergentes[:3]))
    finally:
        arq.unlink(missing_ok=True)

print("\n" + ("TUDO VERDE" if ok else "VERMELHO") + "\n")
sys.exit(0 if ok else 1)
